vit-servicing-station-lib/src/db/queries/search/mod.rs

```diff
@@ -0,0 +1,196 @@
+mod generic;
+
+use crate::{
+    db::{
+        queries::search::generic::{search, RANDOM},
+        DbConnection, DbConnectionPool,
+    },
+    v0::{endpoints::search::requests::*, errors::HandleError},
+};
+use diesel::r2d2::{ConnectionManager, PooledConnection};
+
+pub async fn search_db(
+    SearchRequest {
+        table,
+        column,
+        sort,
+        query,
+        reverse,
+    }: SearchRequest,
+    pool: &DbConnectionPool,
+) -> Result<SearchResponse, HandleError> {
+    let db_conn = pool.get().map_err(HandleError::DatabaseError)?;
+    tokio::task::spawn_blocking(move || {
+        validate_table_column_sort(table, column, sort)?;
+        let mut response = match table {
+            SearchTable::Proposal => search_proposals(column, sort, &query, &db_conn),
+            SearchTable::Challenge => search_challenges(column, sort, &query, &db_conn),
+        }?;
+
+        if reverse {
+            response.reverse();
+        }
+
+        Ok(response)
+    })
+    .await
+    .map_err(|_e| HandleError::InternalError("Error executing request".to_string()))?
+}
+
+fn search_proposals(
+    column: SearchColumn,
+    sort: SearchSort,
+    query: &str,
+    conn: &PooledConnection<ConnectionManager<DbConnection>>,
+) -> Result<SearchResponse, HandleError> {
+    use crate::db::views_schema::full_proposals_info::dsl::*;
+    use full_proposals_info as proposals;
+
+    let vec = match column {
+        SearchColumn::ProposalTitle => match sort {
+            SearchSort::Index => search(proposals, proposal_title, id, query, conn),
+            SearchSort::Random => search(proposals, proposal_title, RANDOM, query, conn),
+            SearchSort::ProposalFunds => {
+                search(proposals, proposal_title, proposal_funds, query, conn)
+            }
+            SearchSort::ProposalAdvisor => {
+                search(proposals, proposal_title, proposer_name, query, conn)
+            }
+            SearchSort::ProposalTitle => {
+                search(proposals, proposal_title, proposal_title, query, conn)
+            }
+            _ => unreachable!(),
```

Comment:
I would just move the error thrown inside `validate_table_column_sort` here and just get rid of that function and these `unreachable!`. Less risk of both functions getting out of sync in the future and reaching a panic. But it's a bit of a personal preference I guess.

Reply:
Yeah this didn't make me feel great, part of me feels like there might be a different way of expressing this, perhaps something like:
```rust
enum SearchQuery {
  Challenges(ChallengeColumn),
  Proposals(ProposalColumn),
}

enum ChallengeColumn {
  Index,
  Title,
  // ...
}
// etc...
```
but I'm not sure how nice the json representation would be :shrug: 

But either way, if we stay with the current representation this seems like a sensible change :+1:

Reply:
Perhaps this is better?
```rust
#[derive(Debug, Clone, Deserialize)]
pub struct Query {
    table: String,
    constraints: Vec<Constraint>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Constraint {
    column: String,
    kind: ConstraintKind,
}

#[derive(Debug, Clone, Deserialize)]
pub enum ConstraintKind {
    Contains(String),
    Range { from: i64, to: i64 },
}
```
This seems more flexible, not sure how hard it would be to implement, but I suspect it's doable. The json representation is also quite nice (with some tweaking)